share/CMakeJS.cmake

```diff
@@ -0,0 +1,640 @@
+endif()
+
+#[=============================================================================[
+Get NodeJS C++ Addon development files.
+
+Provides
+::
+
+  NODE_ADDON_API_DIR, where to find napi.h, etc.
+  NODE_ADDON_API_INC_FILES, the headers required to use Node Addon API.
+
+]=============================================================================]#
+function(cmakejs_acquire_napi_cpp_files)
+    execute_process(
+      COMMAND "${NODE_EXECUTABLE}" -p "require('node-addon-api').include"
+      WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+      OUTPUT_VARIABLE NODE_ADDON_API_DIR
+      # COMMAND_ERROR_IS_FATAL ANY
+    )
+    string(REGEX REPLACE "[\r\n\"]" "" NODE_ADDON_API_DIR "${NODE_ADDON_API_DIR}")
+    set(NODE_ADDON_API_DIR "${NODE_ADDON_API_DIR}" CACHE PATH "Node Addon API Headers directory." FORCE)
+
+    unset(NODE_ADDON_API_INC_FILES)
+    file(GLOB NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_DIR}/*.h")
+    set(NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_INC_FILES}" CACHE FILEPATH "Node Addon API Header files." FORCE)
+    source_group("Node Addon API (C++)" FILES "${NODE_ADDON_API_INC_FILES}")
+
+    if(VERBOSE)
+        message(STATUS "NODE_ADDON_API_DIR: ${NODE_ADDON_API_DIR}")
+    endif()
+endfunction()
+
+# Acquire if needed...
+if(NOT DEFINED NODE_ADDON_API_DIR)
+    cmakejs_acquire_napi_cpp_files()
+    message(STATUS "NODE_ADDON_API_DIR: ${NODE_ADDON_API_DIR}")
+endif()
+if(NOT DEFINED NODE_ADDON_API_INC_FILES)
+    file(GLOB NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_DIR}/*.h")
+    set(NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_INC_FILES}" CACHE FILEPATH "Node Addon API Header files." FORCE)
+    source_group("Node Addon API (C++)" FILES "${NODE_ADDON_API_INC_FILES}")
+endif()
+
+#[=============================================================================[
+Silently create an interface library (no output) with all Addon API dependencies
+resolved, for Addon targets to link with.
+
+(This should contain most of cmake-js globally-required configuration)
+
+Targets:
+
+cmake-js::node-dev
+cmake-js::node-api
+cmake-js::node-addon-api
+cmake-js::cmake-js
+
+]=============================================================================]#
+
+# NodeJS system installation headers
+# cmake-js::node-dev
+add_library                 (node-dev INTERFACE)
+add_library                 (cmake-js::node-dev ALIAS node-dev)
+target_include_directories  (node-dev INTERFACE "${CMAKE_JS_INC}")
+target_sources              (node-dev INTERFACE "${CMAKE_JS_SRC}")
+target_sources              (node-dev INTERFACE "${CMAKE_JS_INC_FILES}")
+target_link_libraries       (node-dev INTERFACE "${CMAKE_JS_LIB}")
+
+# Node API (C) - requires NodeJS system installation headers
+# cmake-js::node-api
+add_library                 (node-api INTERFACE)
+add_library                 (cmake-js::node-api ALIAS node-api)
+target_include_directories  (node-api INTERFACE "${NODE_API_HEADERS_DIR}")
+target_sources              (node-api INTERFACE "${NODE_API_INC_FILES}")
```

Comment:
what is the purpose of this library with the headers as sources? Nothing needs explicit compiling here, so what does this do?

Reply:
[Related](https://github.com/cmake-js/cmake-js/issues/309#issuecomment-1756218850)

[Related](https://github.com/cmake-js/cmake-js/issues/310#issuecomment-1785905574)

This is conventional CMake.

I spotted some discussion and code which posited to restrict one's access to the C API, if I'm not mistaken.

It is not possible to use the C++ API without the C API underneath it, AFAIK.

But I might *want* to use the C API directly, or even ```Node::Stuff()``` and ```v8::Stuff()``` (the node development headers that cmake-js fetches). I mean, it's the underlying code for the whole Addon API, and it isn't really cmake-js' API to deprecate or restrict access to.

If one actually specifically wants to use ```Node::Stuff()``` and ```v8::Stuff()``` then they can just:

```.cmake
add_library(my_nodejs_extending_target SHARED)
target_link_libraries(my_nodejs_extending_target cmake-js::node-dev)
```

And only those headers shall be made available to that target.

Reply:
Being convention is fine, it just seems like pointless work to me :shrug:

> I spotted some discussion and code which posited to restrict one's access to the C API, if I'm not mistaken.

That was specifically to do with the 'full' headers (ie `libnode-dev`), which contain the unstable C API of nodejs (predates node-api). The `node-api-headers` package provides the node-api C API which is ok to be used by user code.

Reply:
All fair. I noted your  'monolothic' approach on the other branch and don't object to it in the initial case, but I'm absolutely certain that the multi-target functionality would be widely requested later on - assuming users were adopting.

I agree there is no benefit in needless complications, but will counter re: this implementation that INTERFACE libs avoid being 'pointless work' by not compiling anything of their own. When a to-be-compiled target links to an interface lib, the interface's contents will end up inside the to-be-compiled's binary output. If nothing links, they technically 'don't exist', other than inside the contents of the ```export()```'s output file. No work is done and they can be safely thought to not exist.

My original draft actually only made a single INTERFACE lib which contained *everything*. I decided to part the libs out logically after noting several such requests, and thinking ahead.

See also comments about vcpkg and exporting/installing of CMake modules below.